### ISSUE TITLE
B0 mapping: Phase unwrapping

### ISSUE THREAD

**Issue by IanGagnon7** [neuropoly/qMRLab#48]
The "prelude" unwrapping function of unix have been replace by a matlab function: https://github.com/neuropoly/qMRLab/commit/1a00c080b6564964dda34212778a6bd5e552af5a
Here is the difference between B0map obtain with both function. As we can see, in the important region, the difference is negligible! So I supposed the new function works well enough to move forward with it. 
*** The main goal of this replacement is to remove the dependence on unix. ***
What do you guys think? @mathieuboudreau @tanguyduval 
![screen shot 2017-07-28 at 2 41 05 pm](https://user-images.githubusercontent.com/28765087/28731680-dfa849d4-73a2-11e7-97e3-4a3d45b918a2.png)

**Comment by tanguyduval** [comment]
remove dependency on FSL.
Plz move license +  sunwrap to external folder.

**Comment by jcohenadad** [comment]
This is not a proper demonstration that your tool produces the same result as fsl. The phantom has not structure and B0 is flat there

**Comment by mathieuboudreau** [comment]
@IanGagnon7 Do you have a set of B0 DEM human data from the scanner to test it on?

**Comment by mathieuboudreau** [comment]
@jcohenadad I think the image that @IanGagnon7 showed was simply the difference between the two unwrapped phase maps of a numerical phantom (which contained both noise and "structure").

If you run the example file [here](https://github.com/neuropoly/qMRLab/blob/1a00c080b6564964dda34212778a6bd5e552af5a/Models_Functions/B0_DEMfun/example.m), it loads a [mat file](https://github.com/neuropoly/qMRLab/blob/1a00c080b6564964dda34212778a6bd5e552af5a/Models_Functions/B0_DEMfun/example.mat) that contains data with varying noise and structure, as can be seen from the output of example.m (below).

![example_result](https://user-images.githubusercontent.com/1421029/28745453-91dfae3e-7453-11e7-8747-d3902ee546c8.png)

Is this correct, @IanGagnon7 ? Where did you get this sample simulated data from?

**Comment by mathieuboudreau** [comment]
I asked my colleague Yuhan Ma who has a lot of experience with phase data if she knew of any good dataset to test phase unwrapping on, and she pointed me to the QSM recon challenge dataset: http://www.neuroimaging.at/pages/qsm.php

I don't think it has B0 mapping data, but it has a 3D brain phase dataset (both raw and unwrapped) that you could test the phase unwrapping tools on. If you download `qsm_recon_challenge_2016.zip` from that page, the relevant *.mat files would be `data/phs_wrap.mat` and `data/phs_unwrap.mat`.

A+

**Comment by tanguyduval** [comment]
Very relevant, thank you mathieu. Ian, you should definitely try on this dataset and post the result here! thx

**Comment by IanGagnon7** [comment]
I agree with Tanguy, thanks for your pertinent comment @mathieuboudreau !
I will test both techniques tomorrow morning with the dataset and I will post the results.
Ian

**Comment by IanGagnon7** [comment]
Unwrapping phase test with QSM recon challenge dataset both for FSL "prelude" and MATLAB "sunwrap".
![screen shot 2017-08-01 at 4 45 46 pm](https://user-images.githubusercontent.com/28765087/28846309-eb680686-76d8-11e7-9c5a-51157bf55115.png)
And also the error for each (difference with the ground truth) and the difference between them.
![screen shot 2017-08-01 at 4 56 05 pm](https://user-images.githubusercontent.com/28765087/28846692-5793de88-76da-11e7-9439-1632bae47f94.png)
As we can see, these two unwrapping function (FSL "prelude" and MATLAB "sunwrap") gave almost the same result (if we look at the difference between them which is very small).
The big issue here is why the error with respect to the ground truth is so high (on average between 3 and 4)?
Any ideas ?

Here is my code!

```
% Phase unwrapping comparison
% FSL "prelude"  vs  MATLAB "sunwrap"
% Run this script section by section
% Written by: Ian Gagnon, 2017

%**************************************************************************
%% FSL "prelude"
%**************************************************************************
unix('prelude -p phs_wrap.nii.gz -a magn.nii.gz -o Ph_uw -f')
B0 = load_untouch_nii('Ph_uw.nii.gz');
B0.img = UNWRAP(B0.img,[],4);
FSL = B0.img(:,:,80);
FSL(FSL==0) = nan;

% figure 1
figure, imagesc(FSL), title('FSL'), colorbar;

%**************************************************************************
%% MATLAB "sunwrap"
%**************************************************************************
tmp1 = load('phs_wrap.mat');
Phase = tmp1.phs_wrap;
tmp2 = load('magn.mat');
Magn = tmp2.magn;
Complex = Magn.*exp(Phase*1i);
MATLAB = sunwrap(Complex(:,:,80));
MATLAB(MATLAB==0) = nan;
MATLAB = MATLAB - 2*pi;

% figure 2
figure, imagesc(MATLAB), title('MATLAB'), colorbar;

%**************************************************************************
%% REF
%**************************************************************************

% figure 3
RAW_PHASE = phs_wrap(:,:,80);
figure, imagesc(RAW_PHASE), title('RAW PHASE'), colorbar;

% figure 4
GROUND_TRUTH = phs_unwrap(:,:,80);
GROUND_TRUTH(GROUND_TRUTH==0) = nan;
figure, imagesc(GROUND_TRUTH), title('GROUND TRUTH'), colorbar;

%**************************************************************************
%% Unwrapping results
%**************************************************************************
figure
subplot(2,3,2), imagesc(RAW_PHASE), title('RAW PHASE'), colorbar;
subplot(2,3,4), imagesc(GROUND_TRUTH,[-10 10]), title('Unwrapped phase - GROUND TRUTH'), colorbar;
subplot(2,3,5), imagesc(FSL,[-10 10]), title('Unwrapped phase - FSL'), colorbar;
subplot(2,3,6), imagesc(MATLAB,[-10 10]), title('Unwrapped phase - MATLAB'), colorbar;

%**************************************************************************
%% Comparison
%**************************************************************************
figure
subplot(1,3,1), imagesc(GROUND_TRUTH - FSL,[-3 8]), title('Difference (GROUND TRUTH - FSL) in radian'), colorbar;
subplot(1,3,2), imagesc(GROUND_TRUTH - MATLAB,[-3 8]), title('Difference (GROUND TRUTH - MATLAB) in radian'), colorbar;
subplot(1,3,3), imagesc(FSL - MATLAB, [-1e-5 1e-5]), title('Difference (GROUND TRUTH - FSL) in radian'), colorbar;
```

Best,
Ian

**Comment by mathieuboudreau** [comment]
It's encouraging to see that `sunwrap` seems to give the same results as `prelude`!

As for the difference between them and the ground truth data, I noticed that you only unwrapped a single slice (axial, slice #80); it's likely that the ground truth data unwrapped the entire volume.

If there is phase wrapping in the ventral->dorsal direction as well, this could explain some of the difference you see between the ground truth and your single-slice results.

**Comment by tanguyduval** [comment]
Thx Mathieu for the feedback. @IanGagnon7 test on the 3D dataset please! sunwrap is works on 3D also.
Should we add the ref in the Class Definition?
````
Florian Maier, David Fuentes, Jeffrey S. Weinberg, John D. Hazle, and R. Jason Stafford
Robust Phase Unwrapping for MR Temperature Imaging Using a Magnitude-Sorted List, Multi-Clustering Algorithm
Magnetic Resonance in Medicine, 73(4):1662–1668, 2015.
````

**Comment by IanGagnon7** [comment]
FSL `prelude` function already unwrapped the entire volume, I simply showed one slice. But it's a good point for MATLAB `sunwrap`, when unwrapping the entire volume, some problems occurred...
![screen shot 2017-08-02 at 3 32 41 pm](https://user-images.githubusercontent.com/28765087/28891251-dfbcae94-7797-11e7-92c6-9e3824676880.png)

EARLY CONCLUSION: 
- In 2D, `sunwrap` reproduced `prelude` fairly accurately
- In 3D, `sunwrap` seems to have some issues while `prelude` works as well as in 2D

Because `prelude` is globally better than `sunwrap`, I propose to keep both and give the option of using one or the other. If the user has FSL, `prelude` would be the one, if not, he could still uses `sunwrap`.
What do you guys think? @mathieuboudreau  @tanguyduval 

PS: @mathieuboudreau  Your comment/feedback are always very relevant and helpful.
A big thanks for taking the time to answer me!!

**Comment by mathieuboudreau** [comment]
That's a weird-looking artifact!

I called Yuhan and we looked over the paper; it looks like this artifact may be due to the fact that this `sunwrap` uses clusters of voxels during the phase unwrapping. I've only glanced at it, but it sounds like they are sorting the voxels by SNR, and then they are creating clusters through the neighbours of the highest SNR voxels until the clusters meet neighbouring clusters. It makes sense that an artifact could be present in the 3D data and not 2D, since [3D looks at the 26 neighbouring voxels instead of just 8 for the 2D data](https://github.com/fmaier/MRM-2014-PhaseUnwrapping/blob/master/sunwrap.m#L127-L128). One other funny thing I see in the data is that `sunwrap` has some phase surrounding the cortex that looks a bit wacky as well.

You could raise an issue to [sunwrap's GitHub repo](https://github.com/fmaier/MRM-2014-PhaseUnwrapping) to point out this artifact on this dataset, maybe the authors have already implemented a fix internally and forgot to share it? Or you could email the authors; they should have an email listed as their contact info on their paper.

Lastly, Yuhan pointed me to another possible phase unwrapping tool implemented in Matlab that you could try, called `laplacianUnwrap.m`, and is part of a QSM tooldbox on the Matlab File Exchange: https://www.mathworks.com/matlabcentral/fileexchange/48557-quantitative-susceptibility-mapping-for-mri-part-1?focused=6248679&tab=function

A+

p.s. No problem! Glad I could be of some help!

**Comment by tanguyduval** [comment]
amazing feedback mathieu, thanks a lot again.
@IanGagnon7 try with laplacianUnwrap. When you compare with the ground truth display between -pi and pi (as opposed to -10 and 10).
Thx

**Comment by IanGagnon7** [comment]
I'd test `laplacianUnwrap` and here are the results:
Note: The display is between -5 and 5, because to much information was lost between -pi and pi
![screen shot 2017-08-03 at 4 28 53 pm](https://user-images.githubusercontent.com/28765087/28942315-e82fe81c-7868-11e7-86f1-65b88397cc08.png)


EARLY CONCLUSION:
- In 2D, `laplacianUnwrap` doesn't seem to work at all
- In 3D, `laplacianUnwrap` works very well, even better than FSL `prelude`

We could take `laplacianUnwrap` for the 3D and `sunwrap` for the 2D ? 
But it's not really professional to have 2 different algorithms for unwrapping  2D and 3D... 
In the light of all these results, FSL `prelude` stays to best unwrapping function we have found so far. As I propose in my last post, we could simply keep both `prelude` and `sunwrap`. That will allows user who has FSL to use `prelude` and user who hasn't to use `sunwrap`.

Best,
Ian

**Comment by tanguyduval** [comment]
Thanks Ian, for now use laplacianUnwrap for the 3D and sunwrap for the 2D .
Best

**Comment by mathieuboudreau** [comment]
I agree with Tanguy, let's do that until we find one that does both cases.

**Comment by rtopfer** [comment]
Comparing phase unwrapping techniques using 1) simulated phase maps 2) in vivo data:
 
**1) Simulation phase maps**
<img width="794" alt="screen shot 2018-02-26 at 10 55 25 pm" src="https://user-images.githubusercontent.com/3819344/36709857-38bd4d60-1b48-11e8-81ed-9a006587f831.png">

<img width="1267" alt="screen shot 2018-02-26 at 10 55 35 pm" src="https://user-images.githubusercontent.com/3819344/36709862-401a6bc4-1b48-11e8-88fd-ad07b851fad2.png">

**2) Field mapping (in vivo)**
<img width="790" alt="screen shot 2018-02-26 at 10 55 44 pm" src="https://user-images.githubusercontent.com/3819344/36709865-45316dc4-1b48-11e8-937c-6ad072f9088a.png">

**Comment by spinicist** [comment]
Hi,

You might be interested in this [paper](https://doi.org/10.1371/journal.pone.0149602) which compared the Abdul-Rahman to Laplacian unwrapping for QSM, and concluded that the Laplacian was not suitable. The Laplacian intrinsically will always alter phase values at least slightly.

@rtopfer - how were you calculating the residuals? The unwrapped phase is a degenerate value mod 2*pi - i.e., if your image is the same as the ground truth +/- a global integer multiple of 2*pi, that's actually a valid solution. If the Abdul-Rahman method picks the "wrong" region as having the highest reliability, then it will produce such an answer. It looks to me like this may have happened. Also, I would argue that a comparison without noise is not very useful. The whole point of the Abdul-Rahman method is that it is *very* robust against noise.

I recently implemented the Abdul-Rahman method in [QUIT](https://github.com/spinicist/QUIT/blob/modules/Source/Susceptibility/qi_unwrap_path.cpp), after finding it in the MEDI toolbox. In my opinion, from running it on a few datasets here, the results are as robust as `prelude`, but it is far faster, so personally I'm planning to use it from now on. I experienced the 2*pi issue above while writing my tests for it.

**Comment by rtopfer** [comment]
Hi,

Indeed, in the simulation I posted above, that's exactly what has happened for the phase map in the 2nd row [0:2]: where the Abdul-Rahman solution is off, it is by a global 2pi offset. This can happen with any of the methods and, in my experience with in vivo data, prelude and the Abdul-Rahman method (which I normally use) generally give very similar results. +Sure, you could add noise to the simulation to make a more realistic comparison between methods, but my specific point was to demonstrate, as in the paper mentioned by @spinicist, that the Laplacian technique doesn't accurately unwrap the phase in the _ideal_ case; a fortiori, it won't yield the correct phase in the noisy real-world scenario either. So, it likely shouldn't be the default unwrapping technique in qMRLab.

**Comment by spinicist** [comment]
@rtopfer - Great, sounds like we agree.

I'm not sure how you calculated your maps, but just in case, there is a MEX version of Abdul-Rahman in the MEDI toolbox (available from Cornell's website), and there's also versions in several languages available on the web-site of the original authors.

**Comment by mathieuboudreau** [comment]
Anyone see this paper? http://onlinelibrary.wiley.com/doi/10.1002/mrm.26989/abstract

Their code is in available online and is written in MATLAB with an MIT license: https://gitlab.com/veronique_fortier/Quality_guided_unwrapping

**Comment by spinicist** [comment]
Thanks, I've not seen that paper before. From a skim read, there method is an adaption of Abdul-Rahman to improve speed. I don't think Abdul-Rahman is particularly slow (my version is far faster than Prelude), so the benefit may be minor.

They also compare it to the code in MEDI, which I thought was also a version of Abdul-Rahman (I don't think there are any comments in the MEX file indicating which algorithm it is, but it looked like an alternate implementation to me. I might have got that totally wrong).
